api/v1beta1/validator.go

```diff
@@ -30,7 +47,121 @@ func validOcid(ocid string) bool {
 }
 
 // validShape is a simple pre-flight
-// we will let the serverside handle the more complex and compete validation
+// we will let the serverside handle the more complex and compete validation.
 func validShape(shape string) bool {
 	return len(shape) > 0
 }
+
+// validRegion test if the string can be a region.
+func validRegion(stringRegion string) bool {
+	if blankRegex.MatchString(stringRegion) || stringRegion == "" {
+		return false
+	}
+	return true
+}
+
+// validateClusterName validates ClusterName.
+func (c *OCICluster) validateClusterName() field.ErrorList {
+	var allErrs field.ErrorList
+	if len(c.Name) > clusterNameMaxLength {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), c.Name,
+			fmt.Sprintf("Cluster Name longer than allowed length of %d characters", clusterNameMaxLength)))
+	}
+
+	if blankRegex.MatchString(c.Name) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), c.Name,
+			fmt.Sprintf("cluster name can not contain spaces")))
+	}
+
+	if c.Name == "" {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), c.Name,
+			fmt.Sprintf("cluster name can not be empty")))
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return allErrs
+}
+
+// validateNetworkSpec validates the NetworkSpec
+func validateNetworkSpec(networkSpec NetworkSpec, old NetworkSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if len(networkSpec.Vcn.CIDR) > 0 {
+		allErrs = append(allErrs, validateVCNCIDR(networkSpec.Vcn.CIDR, fldPath.Child("cidr"))...)
+	}
+
+	if networkSpec.Vcn.Subnets != nil {
+		allErrs = append(allErrs, validateSubnets(networkSpec.Vcn.Subnets, networkSpec.Vcn, fldPath.Child("subnets"))...)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return allErrs
+}
+
+// validateVCNCIDR validates the CIDR of a VNC.
+func validateVCNCIDR(vncCIDR string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if _, _, err := net.ParseCIDR(vncCIDR); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, vncCIDR, "invalid CIDR format"))
+	}
+	return allErrs
+}
+
+// validateSubnetCIDR validates the CIDR blocks of a Subnet.
+func validateSubnetCIDR(subnetCidr string, vcnCidr string, fldPath *field.Path) field.ErrorList {
```

Comment:
should we add validation that in the provided NSG sec list rules, the cidr is correct?